core/src/commands/cloud/secrets/secrets-update.ts

```
@@ -223,17 +133,13 @@ export class SecretsUpdateCommand extends Command<Args, Opts> {
         secretsToUpdateArgs,
         log,
       })
+    }
 
-      if (isUpsert) {
-        // if --upsert is set, check the diff between secrets to update and command args to find out
-        // secrets that do not exist yet and can be created
-        secretsToCreate = getSecretsToCreate(secretsToUpdateArgs, secretsToUpdate)
-      }
-    } else {
-      // update secrets by ids
-      secretsToUpdate = sortBy(allSecrets, "name")
-        .filter((secret) => Object.keys(secretsToUpdateArgs).includes(secret.id))
-        .map((secret) => ({ ...secret, newValue: secretsToUpdateArgs[secret.id] }))
+    let secretsToCreate: [string, string][] = []
+    if (isUpsert && !updateById) {
```

It might be a bit more readable to keep it in the else block where we already know `!updateById` or was there a specific reason to move it out?

---

The reason was to avoid nested and too complex if-conditions. It would be nice to extract initializer methods for both arrays to declare them as `const` instead of `let`.
Moreover, it looks like these 2 flags cannot be combined, I'm gonna test it and add the validation if necessary.
That will be done in the next PR.

Maybe, another way will be to avoid fetching all secrets in advance and splitting the input data to 2 collections. Instead, we can just create secrets one-by-one and apply the necessary behaviour on conflict.

---

Anyway, this will be changed soon, in the next PR :)

This PR just brings some preliminary refactoring to make the further changes in create/update secrets easier.

---

Ah i see. I guess it makes sense that you can't upsert a secret by id because that would allow you to set the id if it does not exist.

---

Yes, that definitely makes sense. But it looks like Garden CLI can accept both flags now and silently ignore `--upsert` if both are provided. That should be documented or validated with error. That's also will be done in the next PR :)